tests/fuzz/s2n_server_cert_recv_test.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/* Target Functions: s2n_server_cert_recv s2n_x509_validator_validate_cert_chain */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+
+#include "api/s2n.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_config.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_crypto.h"
+#include "tls/s2n_tls.h"
+#include "tls/s2n_tls_parameters.h"
+#include "utils/s2n_random.h"
+#include "utils/s2n_safety.h"
+#include "s2n_test.h"
+
+static const uint8_t TLS_VERSIONS[] = {S2N_TLS10, S2N_TLS11, S2N_TLS12};
+
+static void s2n_fuzz_atexit()
+{
+    s2n_cleanup();
+}
+
+int LLVMFuzzerInitialize(const uint8_t *buf, size_t len)
+{
+#ifdef S2N_TEST_IN_FIPS_MODE
+    S2N_TEST_ENTER_FIPS_MODE();
+#endif
+
+    GUARD(s2n_init());
+    GUARD(atexit(s2n_fuzz_atexit));
+    return 0;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *buf, size_t len)
+{
+    for(int version = 0; version < s2n_array_len(TLS_VERSIONS); version++) {
+        /* Setup */
+        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+        notnull_check(conn);
+        conn->actual_protocol_version = TLS_VERSIONS[version];
+        conn->x509_validator.skip_cert_validation = 1;
```

Comment:
I think we should try both options for `skip_cert_validation ` for better code coverage of our cert validation error handling. Can you add another for loop at the top that does that?

Reply:
You're entirely right, enabling it increases coverage, so it's now in the latest version. To optimize our fuzzing, I used a segment of the libfuzzer input to determine whether skip is set instead of a for loop. This gives libfuzzer control of the choice so it can intentionally target specific codepaths, instead of forcing it to explore multiple codepaths during the same run. Realistically the TLS version for loop should also be replaced similarly, but since that's present in all of our fuzz tests it will be part of the overall redesign and optimization of our fuzz tests which I'm currently working on.